client.go

```diff
@@ -0,0 +1,188 @@
+	defaultServerCert = "/data/server.crt"
+)
+
+type authCmdLineArgsType struct {
+	// hostname or address to bootstrap to
+	bootstrapServer string
+	certFile        string
+	certKey         string
+	serverCert      string
+}
+
+func (cred *authCmdLineArgsType) setDefaultKeysAndCerts(clientCert, clientKey,
+	serverCert string) {
+	if cred.certFile == "" {
+		cred.certFile = clientCert
+	}
+	if cred.certKey == "" {
+		cred.certKey = clientKey
+	}
+	if cred.serverCert == "" {
+		cred.serverCert = serverCert
+	}
+}
+
+type authCredsType struct {
+	// Cert+privkey that authenticates this client
+	clientCert tls.Certificate
+	// Trusted server certificates
+	trustedCerts x509.CertPool
+}
+
+type client struct {
+	BaseURL string
+	authCredsType
+	HTTPClient *http.Client
+}
+
+// Client initialization
+
+func initClient(args authCmdLineArgsType) (client, error) {
+	var httpsClient client
+	args.setDefaultKeysAndCerts(defaultCertFile, defaultCertKey, defaultServerCert)
+
+	if err := httpsClient.initServerTrust(args); err != nil {
+		return client{}, err
+	}
+	if err := httpsClient.initClientCert(args); err != nil {
+		return client{}, err
+	}
+
+	tlsConf := tls.Config{
+		RootCAs:      &httpsClient.trustedCerts,
+		Certificates: []tls.Certificate{httpsClient.clientCert},
+	}
+
+	transport := http.Transport{
+		TLSClientConfig: &tlsConf,
+	}
+
+	httpsClient.HTTPClient = &http.Client{
+		Transport: &transport,
+	}
+
+	return httpsClient, nil
+}
+
+func (c *client) initServerTrust(args authCmdLineArgsType) error {
+
+	if args.serverCert == "" {
+		panic("certificate should be replaced with default one")
+	}
+	trustedCerts := *x509.NewCertPool()
+	certPoolAppendCertsFromFile(&trustedCerts, args.serverCert)
+
+	if len(trustedCerts.Subjects()) == 0 {
+		return errorAddingServerCertificateToPool
+	}
+	c.trustedCerts = trustedCerts
+	return nil
+}
+
+func (c *client) initClientCert(args authCmdLineArgsType) error {
+	clientCert, err := tls.LoadX509KeyPair(args.certFile, args.certKey)
+	if err != nil {
+		return errorLoadingClientCertificate
+	}
+	c.clientCert = clientCert
+	return nil
+}
+
+func certPoolAppendCertsFromFile(s *x509.CertPool, f string) bool {
+	cacert, err := ioutil.ReadFile(f)
+	if err != nil {
+		log.Warnln("Error reading certificate file ", err)
+		return false
+	}
+
+	return s.AppendCertsFromPEM(cacert)
+}
+
+// Client request sending and parsing
+
+type clientRequestType struct {
+	reqType string
+	request string
+}
+
+type clientWorker interface {
+	formatRequest() clientRequestType
+	actOnResponse(http.Response, []byte) error
+	getClient() client
+}
+
+func makeJobDone(req clientWorker) error {
+	request := req.formatRequest()
+	client := req.getClient()
+
+	response, data, err := client.sendRequest(request.reqType, request.request)
+	if err != nil {
+		return err
+	}
+
+	return req.actOnResponse(*response, data)
+}
+
+func (c *client) sendRequest(reqType string, request string) (*http.Response, []byte, error) {
```

Comment:
Maybe I am doing something wrong here, but Body is a type `Body io.ReadCloser` and you are supposed to close it once reading is done. It looks like when I am trying to do `defer Body.Close()` the Body content is not accessible outside `sendRequest()` function. That is why I am reading the body and storing it's content in separate place.